lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -1137,7 +1137,104 @@ namespace {
         return false;
       }
     }
-    
+
+    /// \brief Estimate how big is the search space that exhaustivity
+    /// checker needs to cover, based on the total space and information
+    /// from the `switch` statement itself. Some of the easy situations
+    /// like `case .foo(let bar)` don't really contribute to the complexity
+    /// of the search so their sub-space sizes could be excluded from
+    /// consideration.
+    ///
+    /// \param total The total space to check.
+    /// \param covered The space covered by the `case` statements in the switch.
+    ///
+    /// \returns The size of the search space exhastivity checker has to check.
+    size_t estimateSearchSpaceSize(const Space &total, const Space &covered) {
+      switch (PairSwitch(total.getKind(), covered.getKind())) {
+        PAIRCASE(SpaceKind::Type, SpaceKind::Type) : {
+          return total.getType()->isEqual(covered.getType())
+                     ? 0
+                     : total.getSize(TC, DC);
+        }
+        PAIRCASE(SpaceKind::Type, SpaceKind::Disjunct)
+            : PAIRCASE(SpaceKind::Type, SpaceKind::Constructor) : {
```

Comment:
This one's especially wrong. Maybe we should look into removing the macro and just using the helper type.